doc/cephadm/operations.rst

```diff
@@ -734,3 +734,64 @@ Purge ceph daemons from all hosts in the cluster
 
   # For each host:
   cephadm rm-cluster --force --zap-osds --fsid <fsid>
+
+
+Replacing a device
+==================
+
+The ``ceph orch device replace`` command automates the process of replacing the underlying device of an OSD.
+Previously, this process required manual intervention at various stages.
+With this new command, all necessary operations are performed automatically, streamlining the replacement process
+and improving the overall user experience.
+
+.. note:: This only supports LVM-based deployed OSD(s)
+
+.. note:: This will work only if ``ceph-volume lvm list`` sees the corresponding OSD(s)
+
+.. prompt:: bash #
+
+  ceph orch device replace <host> <device-path>
+
+In the case the device being replaced is shared by multiple OSDs (eg: DB/WAL device shared by multiple OSDs), the orchestrator will warn you.
+
+.. prompt:: bash #
+
+  [ceph: root@ceph /]# ceph orch device replace osd-1 /dev/vdd
+
+  Error EINVAL: /dev/vdd is a shared device.
+  Replacing /dev/vdd implies destroying OSD(s): ['0', '1'].
+  Please, *be very careful*, this can be a very dangerous operation.
+  If you know what you are doing, pass --yes-i-really-mean-it
+
+If you know what you are doing, you can go ahead and pass ``--yes-i-really-mean-it``.
+
+.. prompt:: bash #
+
+  [ceph: root@ceph /]# ceph orch device replace osd-1 /dev/vdd --yes-i-really-mean-it
+    Scheduled to destroy osds: ['6', '7', '8'] and mark /dev/vdd as being replaced.
+
+``cephadm`` will make ``ceph-volume`` zap and destroy all related devices and mark the corresponding OSD as ``destroyed`` so the
+different OSD(s) ID(s) will be preserved:
+
+.. prompt:: bash #
+
+  [ceph: root@ceph-1 /]# ceph osd tree
+    ID  CLASS  WEIGHT   TYPE NAME         STATUS     REWEIGHT  PRI-AFF
+    -1         0.97659  root default
+    -3         0.97659      host devel-1
+     0    hdd  0.29300          osd.0     destroyed   1.00000  1.00000
+     1    hdd  0.29300          osd.1     destroyed   1.00000  1.00000
+     2    hdd  0.19530          osd.2            up   1.00000  1.00000
+     3    hdd  0.19530          osd.3            up   1.00000  1.00000
+
+The device being replaced is finally seen as ``being replaced`` preventing ``cephadm`` from redeploying the OSDs too fast:
```

Comment:
> Maybe it might be better to word it like:
> 
> ... preventing `cephadm` from redeploying the OSDs until the device has been physically replaced or the `is-being-replaced` device-marker has been removed manually from the disk.

yeh sure, thanks for that suggestion 👍 

> Also, it would be great to also have a `ceph orch device replace cancel <host> <device-path>` command, in case of operator error or just not wanting to actually physically replace the device yet.

There's already the `--i-really-mean-it` flag but it is only required in case of shared devices.
I'm not sure aborting the removal in the middle of the process is a nice thing.

Perhaps we should always require `--i-really-mean-it` and make it necessary to be passed twice when it's a shared device (for instance, I believe that flag is required to be passed twice for osd pool removal).

@adk3798 Thought on this?

Reply:
I was more thinking of operator error like on big clusters an operator mistakenly using the wrong /dev/path and discovering that after that fact and wanting to bring in the OSDs again. 

we could document the dd to the device, but this is also *very* error prone, cu could do much harm by wrong device paths or writing the wrong numbers in `bs` or `count`. 

So I don't mean to interrupt while it's removing/destroying OSDs, but to able to remove the flag and allow the orchestrator to bring the device back in even though it has not been replaced physically.

Reply:
> So I don't mean to interrupt while it's removing/destroying OSDs, but to able to remove the flag and allow the orchestrator to bring the device back in even though it has not been replaced physically.

ok, that's slightly different indeed. In that case we can certainly add that "cancel" command.

```console
# ceph orch device replace cancel <host> <device-path>
```

I would perhaps just not call it "cancel", that is what confused me and made me think it was for interrupting the removal.

Reply:
yeah, as discussed over slack, `clear` or something similar (`clear_flag`) might be better, I am OK with whatever feels best to you guys as long as we have a clean way to remove the flag again. 

Thank you!

Reply:
as long as this cancel is just removing the flag rather than trying to stop the OSD removal part way through, which seems to be the outcome of this thread, I'm okay with that. I don't have any great ideas for a command name off the top of my head.

Reply:
I implemented the `ceph orch device replace <host> <device> --clear` and documented it.